pkg/tests/end_to_end_tests/multi_tenancy_test.go

```diff
@@ -0,0 +1,580 @@
+		expectedResult = []prompb.TimeSeries{
+			{
+				Labels: []prompb.Label{
+					{Name: model.MetricNameLabelName, Value: "secondMetric"},
+					{Name: "job", Value: "baz"},
+					{Name: "ins", Value: "tag"},
+					{Name: tenancy.TenantLabelKey, Value: "tenant-a"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 2, Value: 2.2},
+					{Timestamp: 3, Value: 2.3},
+					{Timestamp: 4, Value: 2.4},
+				},
+			},
+			{
+				Labels: []prompb.Label{
+					{Name: model.MetricNameLabelName, Value: "secondMetric"},
+					{Name: "job", Value: "baz"},
+					{Name: "ins", Value: "tag"},
+					{Name: tenancy.TenantLabelKey, Value: "tenant-b"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 2, Value: 2.2},
+					{Timestamp: 3, Value: 2.3},
+					{Timestamp: 4, Value: 2.4},
+				},
+			},
+			{
+				Labels: []prompb.Label{
+					{Name: model.MetricNameLabelName, Value: "secondMetric"},
+					{Name: "job", Value: "baz"},
+					{Name: "ins", Value: "tag"},
+					{Name: tenancy.TenantLabelKey, Value: "tenant-c"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 2, Value: 2.2},
+					{Timestamp: 3, Value: 2.3},
+					{Timestamp: 4, Value: 2.4},
+				},
+			},
+		}
+
+		result, err = qr.Query(&prompb.Query{
+			Matchers: []*prompb.LabelMatcher{
+				{
+					Type:  prompb.LabelMatcher_EQ,
+					Name:  model.MetricNameLabelName,
+					Value: "secondMetric",
+				},
+			},
+			StartTimestampMs: 2,
+			EndTimestampMs:   4,
+		})
+		require.NoError(t, err)
+
+		// Verifying result.
+		verifyResults(t, expectedResult, result)
+	})
+}
+
+func TestMultiTenancyWithValidTenants(t *testing.T) {
+	ts, tenants := generateSmallMultiTenantTimeseries()
+	withDB(t, *testDatabase, func(db *pgxpool.Pool, t testing.TB) {
+		// With valid tenants.
+		cfg := tenancy.NewSelectiveTenancyConfig(tenants[:2], false) // valid tenant-a & tenant-b.
+		mt, err := tenancy.NewAuthorizer(cfg)
+		require.NoError(t, err)
+
+		// Ingestion.
+		client, err := pgclient.NewClientWithPool(&pgclient.Config{}, 1, pgxconn.NewPgxConn(db), mt)
+		require.NoError(t, err)
+		defer client.Close()
+
+		wauth := mt.WriteAuthorizer()
+		// Ingest tenant-a.
+		request := ingstr.NewWriteRequestWithTs(copyMetrics(ts))
+		err = wauth.Process(requestWithHeaderTenant(tenants[0]), request)
+		require.NoError(t, err)
+		_, err = client.Ingest(request)
+		require.NoError(t, err)
+
+		// Ingest tenant-b.
+		request = ingstr.NewWriteRequestWithTs(copyMetrics(ts))
+		err = wauth.Process(requestWithHeaderTenant(tenants[1]), request)
+		require.NoError(t, err)
+		_, err = client.Ingest(request)
+		require.NoError(t, err)
+		require.NoError(t, err)
+
+		// Ingest tenant-c.
+		request = ingstr.NewWriteRequestWithTs(copyMetrics(ts))
+		err = wauth.Process(requestWithHeaderTenant(tenants[2]), request)
+		require.Error(t, err)
+		require.Equal(t, err.Error(), "write-authorizer process: authorization error for tenant tenant-c: unauthorized or invalid tenant")
+
+		// Querying.
```

Comment:
hmm, tenant-c is never ingested though, so showing that query results don't contain tenant-c doesn't test the query path properly.